File: examples/interop/c/AdvancedC.scala
package examples.interop.c

import scala.scalanative.unsafe.*

type Person = CStruct2[CString, CInt]

@extern
object AdvancedCLib:
  def print_person(p: Ptr[Person]): Unit = extern
  def have_birthday(p: Ptr[Person]): Unit = extern

@main
def advancedCExample(): Unit =
  Zone:
    val personPtr = alloc[Person]()
    
    // Initialize the struct fields
    !(personPtr.at1) = toCString("Scala Native Hero")
    !(personPtr.at2) = 30
    
    println("Scala says: passing struct to C...")
    AdvancedCLib.print_person(personPtr)
    AdvancedCLib.have_birthday(personPtr)
    
    val updatedAge = !(personPtr.at2)
    println(s"Scala says: reading back updated age from C: $updatedAge")

File: examples/interop/c/advanced_c.c
#include <stdio.h>
#include <string.h>

typedef struct {
    char* name;
    int age;
} Person;

void print_person(Person* p) {
    if (p == NULL) return;
    printf("C says: Person is %s, age %d\n", p->name, p->age);
}

void have_birthday(Person* p) {
    if (p == NULL) return;
    p->age += 1;
    printf("C says: Happy Birthday %s! You are now %d.\n", p->name, p->age);
}
